### DISCUSSION TITLE
Open graph properties are not inherited in children routes metadata

### DISCUSSION THREAD

**Discussion by Arctomachine** [vercel/next.js#58189] [App Router]
### Link to the code that reproduces this issue

https://github.com/Arctomachine/next-metadata-open-graph-bug

### To Reproduce

1. Start server, open dev tools
2. Expand head tag
3. Remember og-related properties
4. Go to about page
5. Look at og-related properties for this page. They all are replaced by single title property. Link to image is also removed.
6. Look at twitter-specific properties. They are mostly the same, but link to image and image sizes are removed too.

![image](https://github.com/vercel/next.js/assets/29041820/5259e7e9-ecb2-4010-bd8f-270e6a251e06)
_Home page has all properties_

![image](https://github.com/vercel/next.js/assets/29041820/e9a64db2-ad31-48bd-a581-3f4d2dc85466)
_What remains on about page_

### Current vs. Expected behavior

Expected all open graph properties to be carried over from root layout metadata and only title to be updated. Instead all open graph properties are replaced by single title, links to image are removed (from twitter too)

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:        
  Platform: win32        
  Arch: x64              
  Version: Windows 10 Pro
Binaries:                
  Node: 20.2.0           
  npm: N/A               
  Yarn: N/A              
  pnpm: N/A              
Relevant Packages:
  next: 14.0.1-canary.0
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.1.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

App Router, Metadata (metadata, generateMetadata, next/head)

**Comment by timneutkens** [comment]
The parent metadata can be accessed and passed along when extending it in a child, you can see that documented here: https://nextjs.org/docs/app/api-reference/functions/generate-metadata

The second parameter of `generateMetadata` is the parent data.

**Reply by Arctomachine** [reply]
@timneutkens  How to set this up correctly if I want to extend entire parent openGraph property and only replace title (and/or other fields)? If I do it like this
```js
const parentMeta = await parent
const parentOG = parentMeta.openGraph
const newOG = parentOG ? { ...parentOG, title } : { title }
```
then typescript gives error about not compatible values
```
Types of property  url  are incompatible.
Type  string | URL | null  is not assignable to type  string | URL | undefined 
Type  null  is not assignable to type  string | URL | undefined 
```
Even with nullish check for openGraph it still tries to pass null somehow

**Reply by DevinWinata** [reply]
`Metadata.openGraph.title` support `TemplateString` type just like `Metadata.title`, but seems like it's not behave the same when I just want to change the title between different page.
I think #57857 could fix this behavior (at least for `title` and `description`) since it will inherit from `title` and `description` for `openGraph` or `twitter`.

Edit 1: tried the canary version, `openGraph.title` and `openGraph.description` is now inherited from `title` and `description` (don't need to set `openGraph.title` and `openGraph.description`). But the same behavior will still occur if you set any property of `openGraph` in a children route metadata. it's better now, but I think it should be inherit everything and just change what is being set in a children route metadata.

**Comment by devjiwonchoi** [comment]
@Arctomachine Try this:

```ts
export async function generateMetadata(
  // ...,
  parent: ResolvingMetadata,
): Promise<Metadata> {
  const { openGraph } = await parent

  return {
    openGraph: {
      ...openGraph,
      // ...
    },
  }
}
```

**Reply by Arctomachine** [reply]
Inherited data is not even compatible with expected data as is. Documentation has nothing to say about this. And yet, even if it is somehow supposed to work, it is not as intuitive as documentation portrays it.
```
TS2322: Type
{} | { determiner?: "" | "a" | "an" | "the" | "auto" | undefined; title: AbsoluteTemplateString; description?: string | undefined; emails?: string[] | undefined; ... 10 more ...; ttl?: number | undefined; } | ... 11 more ... | { ...; }
is not assignable to type  Metadata 
 Type '{ determiner?: "" | "a" | "an" | "the" | "auto" | undefined; title: AbsoluteTemplateString; description?: string | undefined; emails?: string[] | undefined; ... 17 more ...; tags?: string[] | undefined; }' is not assignable to type 'Metadata'.
  Types of property  authors  are incompatible.
   Type  string[] | undefined  is not assignable to type  Author | Author[] | null | undefined 
    Type  string[]  is not assignable to type  Author | Author[] | null | undefined 
     Type  string[]  is not assignable to type  Author[] 
      Type  string  has no properties in common with type  Author 
```

**Reply by devjiwonchoi** [reply]
If so, I'd just go with this:

```ts
export async function generateMetadata(
  // ...,
  parent: ResolvingMetadata,
): Promise<Metadata> {
  const { openGraph } = await parent

  return {
    openGraph: {
      ...(openGraph as Metadata['openGraph']),
      // ...
    },
  }
}
```

**Reply by Arctomachine** [reply]
No, even like this there is still major incompatibility.
@timneutkens you closed original issue, but metadata inheritance is far from streamlined. It should be addressed and improved. Either by massively expanding documentation with detailed workflow on different scenarios and use cases or by making it "just work", but preferably both.
